{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:54:54.801880900Z",
     "start_time": "2023-10-17T10:54:54.351769083Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "GENERAL=\"/users/genomics/marta/BLCA\"\n",
    "GENOMEDIR=\"/genomics/users/marta/genomes\"\n",
    "biomart=pd.read_csv(os.path.join(GENOMEDIR,\"transcript_gene_v41.txt\"), skiprows=1, names=['gene_id','gene_name', 'gene_type'])\n",
    "biomart_old=pd.read_csv(os.path.join(GENOMEDIR,\"transcript_gene.txt\"), skiprows=1, names=['gene_id','transcript_id','gene_name_OLD', 'gene_type'])\n",
    "biomart_old.drop(\"transcript_id\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "raw_candidates = pd.read_csv(os.path.join(GENERAL,\"RNASeq_expression/tumorspecific_3TPM_cohorts.csv\"))\n",
    "raw_candidates = raw_candidates.merge(biomart, on=['gene_type','gene_name'])\n",
    "candidates = raw_candidates"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:54:56.605947588Z",
     "start_time": "2023-10-17T10:54:56.559810987Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Ouspenskaia"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \n0        26           8      153             36.52  ENSG00000214429  False  \n1         2           0       26              6.21  ENSG00000163530   True  \n2         0           1       19              4.53  ENSG00000224516  False  \n3         0           0        3              0.72  ENSG00000229131  False  \n4         0          12       38              9.07  ENSG00000230105  False  \n5         0           0        1              0.24  ENSG00000234464  False  \n6         0           0        9              2.15  ENSG00000247402  False  \n7         0           9       20              4.77  ENSG00000249894  False  \n8         0           0       13              3.10  ENSG00000251550  False  \n9         0           1        3              0.72  ENSG00000253154   True  \n10        0           1        5              1.19  ENSG00000253474   True  \n11        0           0        2              0.48  ENSG00000255689  False  \n12        0           0        0              0.00  ENSG00000260073   True  \n13        0           2        7              1.67  ENSG00000262198  False  \n14        0           0        3              0.72  ENSG00000270495  False  \n15        7           3       79             18.85  ENSG00000083782  False  \n16        0           0        6              1.43  ENSG00000162344   True  \n17        3           2       18              4.30  ENSG00000212901  False  \n18        0           0        6              1.43  ENSG00000225882  False  \n19        0           0        1              0.24  ENSG00000251273  False  \n20        0           6       50             11.93  ENSG00000228437  False  \n21        0           1        5              1.19  ENSG00000228636  False  \n22        0           0        0              0.00  ENSG00000230533  False  \n23        4           3       43             10.26  ENSG00000155495   True  \n24        2           1       24              5.73  ENSG00000224927  False  \n25        0           0       24              5.73  ENSG00000231208  False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ouspenskaia = pd.read_csv(\"/genomics/users/marta/nuORFdb/GSE143263_nuORFdb_v1.0_annotations.csv\")\n",
    "ouspenskaia['gene_id'] = ouspenskaia['geneId'].str.split(\".\", expand=True)[0]\n",
    "ouspenskaia_genes = set(ouspenskaia.gene_id.values.tolist())\n",
    "\n",
    "candidates['nuORF'] = np.where(candidates[\"gene_id\"].isin(ouspenskaia_genes), True, False)\n",
    "candidates"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:00.449311666Z",
     "start_time": "2023-10-17T10:54:59.009887230Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Chong"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \\\n0        26           8      153             36.52  ENSG00000214429  False   \n1         2           0       26              6.21  ENSG00000163530   True   \n2         0           1       19              4.53  ENSG00000224516  False   \n3         0           0        3              0.72  ENSG00000229131  False   \n4         0          12       38              9.07  ENSG00000230105  False   \n5         0           0        1              0.24  ENSG00000234464  False   \n6         0           0        9              2.15  ENSG00000247402  False   \n7         0           9       20              4.77  ENSG00000249894  False   \n8         0           0       13              3.10  ENSG00000251550  False   \n9         0           1        3              0.72  ENSG00000253154   True   \n10        0           1        5              1.19  ENSG00000253474   True   \n11        0           0        2              0.48  ENSG00000255689  False   \n12        0           0        0              0.00  ENSG00000260073   True   \n13        0           2        7              1.67  ENSG00000262198  False   \n14        0           0        3              0.72  ENSG00000270495  False   \n15        7           3       79             18.85  ENSG00000083782  False   \n16        0           0        6              1.43  ENSG00000162344   True   \n17        3           2       18              4.30  ENSG00000212901  False   \n18        0           0        6              1.43  ENSG00000225882  False   \n19        0           0        1              0.24  ENSG00000251273  False   \n20        0           6       50             11.93  ENSG00000228437  False   \n21        0           1        5              1.19  ENSG00000228636  False   \n22        0           0        0              0.00  ENSG00000230533  False   \n23        4           3       43             10.26  ENSG00000155495   True   \n24        2           1       24              5.73  ENSG00000224927  False   \n25        0           0       24              5.73  ENSG00000231208  False   \n\n    Chong  \n0   False  \n1   False  \n2   False  \n3   False  \n4   False  \n5   False  \n6   False  \n7   False  \n8   False  \n9   False  \n10  False  \n11  False  \n12  False  \n13  False  \n14  False  \n15  False  \n16  False  \n17  False  \n18  False  \n19  False  \n20  False  \n21  False  \n22  False  \n23  False  \n24  False  \n25  False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chong = pd.read_csv(\"/genomics/users/marta/Chongetal/Chong_etal_2020_SupData3_41467_2020_14968_MOESM5_ESM.csv\", skiprows=1)\n",
    "chong.rename(columns={'Gene_ID' : 'gene_id'}, inplace=True)\n",
    "chong_genes = set(chong.gene_id.values.tolist())\n",
    "\n",
    "candidates['Chong'] = np.where(candidates[\"gene_id\"].isin(chong_genes), True, False)\n",
    "candidates"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:02.728378075Z",
     "start_time": "2023-10-17T10:55:02.675441130Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## vanHeesch"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \\\n0        26           8      153             36.52  ENSG00000214429  False   \n1         2           0       26              6.21  ENSG00000163530   True   \n2         0           1       19              4.53  ENSG00000224516  False   \n3         0           0        3              0.72  ENSG00000229131  False   \n4         0          12       38              9.07  ENSG00000230105  False   \n5         0           0        1              0.24  ENSG00000234464  False   \n6         0           0        9              2.15  ENSG00000247402  False   \n7         0           9       20              4.77  ENSG00000249894  False   \n8         0           0       13              3.10  ENSG00000251550  False   \n9         0           1        3              0.72  ENSG00000253154   True   \n10        0           1        5              1.19  ENSG00000253474   True   \n11        0           0        2              0.48  ENSG00000255689  False   \n12        0           0        0              0.00  ENSG00000260073   True   \n13        0           2        7              1.67  ENSG00000262198  False   \n14        0           0        3              0.72  ENSG00000270495  False   \n15        7           3       79             18.85  ENSG00000083782  False   \n16        0           0        6              1.43  ENSG00000162344   True   \n17        3           2       18              4.30  ENSG00000212901  False   \n18        0           0        6              1.43  ENSG00000225882  False   \n19        0           0        1              0.24  ENSG00000251273  False   \n20        0           6       50             11.93  ENSG00000228437  False   \n21        0           1        5              1.19  ENSG00000228636  False   \n22        0           0        0              0.00  ENSG00000230533  False   \n23        4           3       43             10.26  ENSG00000155495   True   \n24        2           1       24              5.73  ENSG00000224927  False   \n25        0           0       24              5.73  ENSG00000231208  False   \n\n    Chong  vanHeesch  \n0   False      False  \n1   False      False  \n2   False      False  \n3   False      False  \n4   False      False  \n5   False      False  \n6   False      False  \n7   False      False  \n8   False      False  \n9   False      False  \n10  False      False  \n11  False      False  \n12  False      False  \n13  False      False  \n14  False      False  \n15  False      False  \n16  False      False  \n17  False      False  \n18  False      False  \n19  False      False  \n20  False      False  \n21  False      False  \n22  False      False  \n23  False      False  \n24  False      False  \n25  False      False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n      <th>vanHeesch</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vanHeesch = pd.read_csv(\"/datasets/marta/vanHeesch2023/SuppTable1.csv\", skiprows=1)\n",
    "vanHeesch.rename(columns={'Gene stable ID' : 'gene_id'}, inplace=True)\n",
    "vanHeesch_genes = set(vanHeesch.gene_id.values.tolist())\n",
    "\n",
    "candidates['vanHeesch'] = np.where(candidates[\"gene_id\"].isin(vanHeesch_genes), True, False)\n",
    "candidates"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:03.989576812Z",
     "start_time": "2023-10-17T10:55:03.908438688Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## IEAtlas"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "293222\n"
     ]
    },
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \\\n0        26           8      153             36.52  ENSG00000214429  False   \n1         2           0       26              6.21  ENSG00000163530   True   \n2         0           1       19              4.53  ENSG00000224516  False   \n3         0           0        3              0.72  ENSG00000229131  False   \n4         0          12       38              9.07  ENSG00000230105  False   \n5         0           0        1              0.24  ENSG00000234464  False   \n6         0           0        9              2.15  ENSG00000247402  False   \n7         0           9       20              4.77  ENSG00000249894  False   \n8         0           0       13              3.10  ENSG00000251550  False   \n9         0           1        3              0.72  ENSG00000253154   True   \n10        0           1        5              1.19  ENSG00000253474   True   \n11        0           0        2              0.48  ENSG00000255689  False   \n12        0           0        0              0.00  ENSG00000260073   True   \n13        0           2        7              1.67  ENSG00000262198  False   \n14        0           0        3              0.72  ENSG00000270495  False   \n15        7           3       79             18.85  ENSG00000083782  False   \n16        0           0        6              1.43  ENSG00000162344   True   \n17        3           2       18              4.30  ENSG00000212901  False   \n18        0           0        6              1.43  ENSG00000225882  False   \n19        0           0        1              0.24  ENSG00000251273  False   \n20        0           6       50             11.93  ENSG00000228437  False   \n21        0           1        5              1.19  ENSG00000228636  False   \n22        0           0        0              0.00  ENSG00000230533  False   \n23        4           3       43             10.26  ENSG00000155495   True   \n24        2           1       24              5.73  ENSG00000224927  False   \n25        0           0       24              5.73  ENSG00000231208  False   \n\n    Chong  vanHeesch  IEAtlas  \n0   False      False    False  \n1   False      False    False  \n2   False      False    False  \n3   False      False    False  \n4   False      False     True  \n5   False      False    False  \n6   False      False    False  \n7   False      False    False  \n8   False      False    False  \n9   False      False     True  \n10  False      False    False  \n11  False      False    False  \n12  False      False     True  \n13  False      False    False  \n14  False      False    False  \n15  False      False    False  \n16  False      False     True  \n17  False      False    False  \n18  False      False    False  \n19  False      False     True  \n20  False      False     True  \n21  False      False    False  \n22  False      False     True  \n23  False      False    False  \n24  False      False    False  \n25  False      False    False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n      <th>vanHeesch</th>\n      <th>IEAtlas</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epitopes_IEAtlas = pd.read_csv(\"/genomics/users/marta/IEAtlas/epitopes_in_cancer_tissues_IEAtlas.txt\", sep=\"\\t\")\n",
    "print(len(epitopes_IEAtlas))\n",
    "basic_info_IEAtlas = pd.read_csv(\"/genomics/users/marta/IEAtlas/epitopes_basic_information_IEAtlas.txt\", sep=\"\\t\")\n",
    "\n",
    "IEAtlas = epitopes_IEAtlas.merge(basic_info_IEAtlas, on=[\"Sequence\", \"Length\", \"ORF_ID\"])\n",
    "IEAtlas_genes = set(IEAtlas.geneId.values.tolist())\n",
    "\n",
    "candidates['IEAtlas'] = np.where(candidates[\"gene_id\"].isin(IEAtlas_genes), True, False)\n",
    "candidates\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:05.770685913Z",
     "start_time": "2023-10-17T10:55:04.824219924Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Present in HCC and predicted SB ?"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [],
   "source": [
    "projects=['liver_adjacent_totalRNA',\"hcc_normal_totalRNA\",\"GSE193567\",\"LIHC_TCGA\"]"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:05.780599810Z",
     "start_time": "2023-10-17T10:55:05.775145006Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3993297/2963460755.py:3: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  SBs = pd.read_csv(os.path.join(\"/projects_eg/projects/marta\",proj,\"analysis/11_netMHCpan_patientspecific/TPM05/all_potential_epitopes_noredundant_allinformation_SB_RiboSeq_pcandmutexpr_TRUE.csv\"))\n"
     ]
    },
    {
     "data": {
      "text/plain": "        transcript_id     SB RiboSeq\n0     ENST00000345080   True    True\n1     ENST00000666716   True   False\n2     ENST00000446720   True   False\n3     ENST00000667774   True   False\n4     ENST00000439694   True   False\n...               ...    ...     ...\n7913  chr10_100983997  False    True\n7914  chr1_62162934_A  False    True\n7915  chr19_35746748_  False    True\n7916  ENST00000435908  False   False\n7917  ENST00000603244  False   False\n\n[7918 rows x 3 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>transcript_id</th>\n      <th>SB</th>\n      <th>RiboSeq</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ENST00000345080</td>\n      <td>True</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ENST00000666716</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENST00000446720</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENST00000667774</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENST00000439694</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>7913</th>\n      <td>chr10_100983997</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>7914</th>\n      <td>chr1_62162934_A</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>7915</th>\n      <td>chr19_35746748_</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>7916</th>\n      <td>ENST00000435908</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7917</th>\n      <td>ENST00000603244</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n<p>7918 rows  3 columns</p>\n</div>"
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SBs_all = pd.DataFrame(columns=['transcript_id','SB','RiboSeq'])\n",
    "for proj in projects:\n",
    "    SBs = pd.read_csv(os.path.join(\"/projects_eg/projects/marta\",proj,\"analysis/11_netMHCpan_patientspecific/TPM05/all_potential_epitopes_noredundant_allinformation_SB_RiboSeq_pcandmutexpr_TRUE.csv\"))\n",
    "    SBs = SBs[['ID','SB','RiboSeq']]\n",
    "    SBs.rename(columns={'ID':'transcript_id'}, inplace=True)\n",
    "    SBs_all = pd.concat([SBs_all, SBs])\n",
    "    # transcripts.extend(SBs.ID.values.tolist())\n",
    "\n",
    "# transcripts = set(transcripts)\n",
    "# transcripts_df = pd.DataFrame(transcripts, columns=['transcript_id'])\n",
    "# transcripts_df\n",
    "SBs_all.drop_duplicates(inplace=True)\n",
    "\n",
    "# if repeated and different value for SB, keep True\n",
    "only_SB = SBs_all.drop('RiboSeq', axis=1).drop_duplicates()\n",
    "only_SB = only_SB.sort_values(by='SB', ascending=False)\n",
    "only_SB = only_SB.drop_duplicates(subset='transcript_id', keep='first')\n",
    "\n",
    "# if repeated and different value for SB, keep True\n",
    "only_RiboSeq = SBs_all.drop('SB', axis=1).drop_duplicates()\n",
    "only_RiboSeq = only_RiboSeq.sort_values(by='RiboSeq', ascending=False)\n",
    "only_RiboSeq = only_RiboSeq.drop_duplicates(subset='transcript_id', keep='first')\n",
    "\n",
    "\n",
    "SBs = only_SB.merge(only_RiboSeq, on=\"transcript_id\")\n",
    "SBs.drop_duplicates(inplace=True)\n",
    "SBs"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:11.367997776Z",
     "start_time": "2023-10-17T10:55:06.571249959Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [
    "biomart_basic = pd.read_csv(\"/genomics/users/marta/genomes/transcript_gene.txt\")\n",
    "biomart_basic.columns=['gene_id','transcript_id','gene_name','gene_type']\n",
    "\n",
    "biomart_basic = biomart_basic[['gene_id','transcript_id']]"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:11.572738429Z",
     "start_time": "2023-10-17T10:55:11.366544994Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [],
   "source": [
    "transcripts_genes_SB = SBs.merge(biomart_basic, on=\"transcript_id\")\n",
    "transcripts_genes_SB.drop(\"transcript_id\", axis=1, inplace=True)\n",
    "transcripts_genes_SB.columns = ['HCC_predicted_SB','HCC_RiboSeq','gene_id']"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:11.659836277Z",
     "start_time": "2023-10-17T10:55:11.618495183Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \\\n0        26           8      153             36.52  ENSG00000214429  False   \n1         2           0       26              6.21  ENSG00000163530   True   \n2         0           1       19              4.53  ENSG00000224516  False   \n3         0           0        3              0.72  ENSG00000229131  False   \n4         0          12       38              9.07  ENSG00000230105  False   \n5         0           0        1              0.24  ENSG00000234464  False   \n6         0           0        9              2.15  ENSG00000247402  False   \n7         0           9       20              4.77  ENSG00000249894  False   \n8         0           0       13              3.10  ENSG00000251550  False   \n9         0           1        3              0.72  ENSG00000253154   True   \n10        0           1        5              1.19  ENSG00000253474   True   \n11        0           0        2              0.48  ENSG00000255689  False   \n12        0           0        0              0.00  ENSG00000260073   True   \n13        0           2        7              1.67  ENSG00000262198  False   \n14        0           0        3              0.72  ENSG00000270495  False   \n15        7           3       79             18.85  ENSG00000083782  False   \n16        0           0        6              1.43  ENSG00000162344   True   \n17        3           2       18              4.30  ENSG00000212901  False   \n18        0           0        6              1.43  ENSG00000225882  False   \n19        0           0        1              0.24  ENSG00000251273  False   \n20        0           6       50             11.93  ENSG00000228437  False   \n21        0           1        5              1.19  ENSG00000228636  False   \n22        0           0        0              0.00  ENSG00000230533  False   \n23        4           3       43             10.26  ENSG00000155495   True   \n24        2           1       24              5.73  ENSG00000224927  False   \n25        0           0       24              5.73  ENSG00000231208  False   \n\n    Chong  vanHeesch  IEAtlas  \n0   False      False    False  \n1   False      False    False  \n2   False      False    False  \n3   False      False    False  \n4   False      False     True  \n5   False      False    False  \n6   False      False    False  \n7   False      False    False  \n8   False      False    False  \n9   False      False     True  \n10  False      False    False  \n11  False      False    False  \n12  False      False     True  \n13  False      False    False  \n14  False      False    False  \n15  False      False    False  \n16  False      False     True  \n17  False      False    False  \n18  False      False    False  \n19  False      False     True  \n20  False      False     True  \n21  False      False    False  \n22  False      False     True  \n23  False      False    False  \n24  False      False    False  \n25  False      False    False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n      <th>vanHeesch</th>\n      <th>IEAtlas</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:11.707655624Z",
     "start_time": "2023-10-17T10:55:11.660437083Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name             gene_type  UC-GENOME  IMvigor210  SNY-2017  \\\n0            CYCSP6  processed_pseudogene         28          85         6   \n1             DPPA2        protein_coding          3           6        15   \n2   ENSG00000224516                lncRNA          1          17         0   \n3   ENSG00000229131                lncRNA          3           0         0   \n4   ENSG00000230105                lncRNA         26           0         0   \n5   ENSG00000234464                lncRNA          0           0         1   \n6   ENSG00000247402                lncRNA          9           0         0   \n7   ENSG00000249894                lncRNA         11           0         0   \n8   ENSG00000251550  processed_pseudogene          4           7         2   \n9   ENSG00000253154                lncRNA          2           0         0   \n10  ENSG00000253474                lncRNA          4           0         0   \n11  ENSG00000255689                lncRNA          2           0         0   \n12  ENSG00000260073                lncRNA          0           0         0   \n13  ENSG00000262198                lncRNA          4           1         0   \n14  ENSG00000270495  processed_pseudogene          3           0         0   \n15             EPYC        protein_coding         10          53         6   \n16            FGF19        protein_coding          4           0         2   \n17         KRTAP3-1        protein_coding          4           4         5   \n18        LINC01456                lncRNA          6           0         0   \n19        LINC02228                lncRNA          1           0         0   \n20        LINC02474                lncRNA         22          19         3   \n21        LINC02663                lncRNA          4           0         0   \n22        LINC03004                lncRNA          0           0         0   \n23           MAGEC1        protein_coding          1          23        12   \n24        NDUFA5P10  processed_pseudogene         17           0         4   \n25       ZBTB46-AS1                lncRNA         21           0         3   \n\n    UNC-108  HdM-BLCA-1  total_n  total_percentage          gene_id  nuORF  \\\n0        26           8      153             36.52  ENSG00000214429  False   \n1         2           0       26              6.21  ENSG00000163530   True   \n2         0           1       19              4.53  ENSG00000224516  False   \n3         0           0        3              0.72  ENSG00000229131  False   \n4         0          12       38              9.07  ENSG00000230105  False   \n5         0           0        1              0.24  ENSG00000234464  False   \n6         0           0        9              2.15  ENSG00000247402  False   \n7         0           9       20              4.77  ENSG00000249894  False   \n8         0           0       13              3.10  ENSG00000251550  False   \n9         0           1        3              0.72  ENSG00000253154   True   \n10        0           1        5              1.19  ENSG00000253474   True   \n11        0           0        2              0.48  ENSG00000255689  False   \n12        0           0        0              0.00  ENSG00000260073   True   \n13        0           2        7              1.67  ENSG00000262198  False   \n14        0           0        3              0.72  ENSG00000270495  False   \n15        7           3       79             18.85  ENSG00000083782  False   \n16        0           0        6              1.43  ENSG00000162344   True   \n17        3           2       18              4.30  ENSG00000212901  False   \n18        0           0        6              1.43  ENSG00000225882  False   \n19        0           0        1              0.24  ENSG00000251273  False   \n20        0           6       50             11.93  ENSG00000228437  False   \n21        0           1        5              1.19  ENSG00000228636  False   \n22        0           0        0              0.00  ENSG00000230533  False   \n23        4           3       43             10.26  ENSG00000155495   True   \n24        2           1       24              5.73  ENSG00000224927  False   \n25        0           0       24              5.73  ENSG00000231208  False   \n\n    Chong  vanHeesch  IEAtlas HCC_predicted_SB HCC_RiboSeq  \n0   False      False    False             True       False  \n1   False      False    False             True        True  \n2   False      False    False              NaN         NaN  \n3   False      False    False             True       False  \n4   False      False     True              NaN         NaN  \n5   False      False    False             True       False  \n6   False      False    False             True       False  \n7   False      False    False              NaN         NaN  \n8   False      False    False             True       False  \n9   False      False     True              NaN         NaN  \n10  False      False    False              NaN         NaN  \n11  False      False    False             True        True  \n12  False      False     True              NaN         NaN  \n13  False      False    False             True       False  \n14  False      False    False             True       False  \n15  False      False    False              NaN         NaN  \n16  False      False     True             True        True  \n17  False      False    False              NaN         NaN  \n18  False      False    False             True       False  \n19  False      False     True             True       False  \n20  False      False     True              NaN         NaN  \n21  False      False    False             True       False  \n22  False      False     True             True       False  \n23  False      False    False             True        True  \n24  False      False    False              NaN         NaN  \n25  False      False    False             True       False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n      <th>vanHeesch</th>\n      <th>IEAtlas</th>\n      <th>HCC_predicted_SB</th>\n      <th>HCC_RiboSeq</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>ENSG00000214429</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>ENSG00000224516</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>lncRNA</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000229131</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000234464</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>lncRNA</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>ENSG00000247402</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>lncRNA</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>ENSG00000249894</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>processed_pseudogene</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>ENSG00000251550</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>lncRNA</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>ENSG00000255689</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>ENSG00000262198</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>processed_pseudogene</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>ENSG00000270495</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>ENSG00000083782</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>ENSG00000212901</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>ENSG00000225882</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>ENSG00000228636</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000224927</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>ENSG00000231208</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates_HCC = candidates.merge(transcripts_genes_SB, on=\"gene_id\", how=\"left\")\n",
    "candidates_HCC"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:11.753562695Z",
     "start_time": "2023-10-17T10:55:11.703894163Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "outputs": [],
   "source": [
    "candidates.to_csv(\"/users/genomics/marta/BLCA/RNASeq_expression/tumorspecific_3TPM_cohorts_10percent_immuno.csv\", index=False)\n",
    "candidates_HCC.to_csv(\"/users/genomics/marta/BLCA/RNASeq_expression/tumorspecific_3TPM_cohorts_10percent_riboseq.csv\", index=False)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T13:15:23.455532698Z",
     "start_time": "2023-10-17T13:15:23.414178739Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "outputs": [],
   "source": [
    "candidates_subset = candidates.drop(['UC-GENOME','IMvigor210',\"SNY-2017\",\"UNC-108\",\"HdM-BLCA-1\",\"total_n\",\"total_percentage\",\"Chong\",\"vanHeesch\"], axis=1)\n",
    "candidates_subset\n",
    "# Create a new column based on the specified conditions\n",
    "candidates_subset['Immunoevidence_Source'] = 'False'  # Initialize the new column with 'False'\n",
    "\n",
    "# Apply conditions to update the new column\n",
    "candidates_subset.loc[(candidates_subset['nuORF'] & candidates_subset['IEAtlas']), 'Immunoevidence_Source'] = 'nuORFdb&IEAtlas'\n",
    "candidates_subset.loc[(candidates_subset['nuORF']) & (~candidates_subset['IEAtlas']), 'Immunoevidence_Source'] = 'nuORFdb'\n",
    "candidates_subset.loc[(~candidates_subset['nuORF']) & (candidates_subset['IEAtlas']), 'Immunoevidence_Source'] = 'IEAtlas'\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:17.687282397Z",
     "start_time": "2023-10-17T10:55:17.674766642Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "outputs": [],
   "source": [
    "candidates_subset_biomartOLD = candidates_subset.merge(biomart_old, on=[\"gene_id\",\"gene_type\"], how=\"left\")\n",
    "candidates_subset_biomartOLD.drop_duplicates(inplace=True)\n",
    "candidates_subset_biomartOLD\n",
    "candidates_subset_biomartOLD.drop([\"nuORF\",\"IEAtlas\",\"gene_id\"], axis=1).to_csv(\"/users/genomics/marta/BLCA/RNASeq_expression/tumorspecific_3TPM_cohorts_10percent_immunoevicenceSource_geneNames.csv\", index=False)\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:19.671439545Z",
     "start_time": "2023-10-17T10:55:19.613296257Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name       gene_type          gene_id  nuORF  IEAtlas  \\\n1             DPPA2  protein_coding  ENSG00000163530   True    False   \n5   ENSG00000230105          lncRNA  ENSG00000230105  False     True   \n11  ENSG00000253154          lncRNA  ENSG00000253154   True     True   \n14  ENSG00000253474          lncRNA  ENSG00000253474   True    False   \n26  ENSG00000260073          lncRNA  ENSG00000260073   True     True   \n32            FGF19  protein_coding  ENSG00000162344   True     True   \n36        LINC02228          lncRNA  ENSG00000251273  False     True   \n41        LINC02474          lncRNA  ENSG00000228437  False     True   \n46        LINC03004          lncRNA  ENSG00000230533  False     True   \n55           MAGEC1  protein_coding  ENSG00000155495   True    False   \n\n   Immunoevidence_Source gene_name_OLD  \n1                nuORFdb         DPPA2  \n5                IEAtlas    AL354793.1  \n11       nuORFdb&IEAtlas    AC100801.1  \n14               nuORFdb    AC044893.1  \n26       nuORFdb&IEAtlas    AC023824.1  \n32       nuORFdb&IEAtlas         FGF19  \n36               IEAtlas     LINC02228  \n41               IEAtlas     LINC02474  \n46               IEAtlas    AL356234.2  \n55               nuORFdb        MAGEC1  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_type</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>IEAtlas</th>\n      <th>Immunoevidence_Source</th>\n      <th>gene_name_OLD</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>ENSG00000163530</td>\n      <td>True</td>\n      <td>False</td>\n      <td>nuORFdb</td>\n      <td>DPPA2</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000230105</td>\n      <td>lncRNA</td>\n      <td>ENSG00000230105</td>\n      <td>False</td>\n      <td>True</td>\n      <td>IEAtlas</td>\n      <td>AL354793.1</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000253154</td>\n      <td>lncRNA</td>\n      <td>ENSG00000253154</td>\n      <td>True</td>\n      <td>True</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>AC100801.1</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000253474</td>\n      <td>lncRNA</td>\n      <td>ENSG00000253474</td>\n      <td>True</td>\n      <td>False</td>\n      <td>nuORFdb</td>\n      <td>AC044893.1</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>ENSG00000260073</td>\n      <td>lncRNA</td>\n      <td>ENSG00000260073</td>\n      <td>True</td>\n      <td>True</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>AC023824.1</td>\n    </tr>\n    <tr>\n      <th>32</th>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>ENSG00000162344</td>\n      <td>True</td>\n      <td>True</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>FGF19</td>\n    </tr>\n    <tr>\n      <th>36</th>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>ENSG00000251273</td>\n      <td>False</td>\n      <td>True</td>\n      <td>IEAtlas</td>\n      <td>LINC02228</td>\n    </tr>\n    <tr>\n      <th>41</th>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>ENSG00000228437</td>\n      <td>False</td>\n      <td>True</td>\n      <td>IEAtlas</td>\n      <td>LINC02474</td>\n    </tr>\n    <tr>\n      <th>46</th>\n      <td>LINC03004</td>\n      <td>lncRNA</td>\n      <td>ENSG00000230533</td>\n      <td>False</td>\n      <td>True</td>\n      <td>IEAtlas</td>\n      <td>AL356234.2</td>\n    </tr>\n    <tr>\n      <th>55</th>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>ENSG00000155495</td>\n      <td>True</td>\n      <td>False</td>\n      <td>nuORFdb</td>\n      <td>MAGEC1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates_subset_biomartOLD[candidates_subset_biomartOLD['Immunoevidence_Source'] != \"False\"]"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:20.756905364Z",
     "start_time": "2023-10-17T10:55:20.750611172Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "outputs": [
    {
     "data": {
      "text/plain": "              gene_type Immunoevidence_Source  gene_name  gene_id  nuORF  \\\n0                lncRNA                 False         10       10     10   \n1                lncRNA               IEAtlas          4        4      4   \n2                lncRNA               nuORFdb          1        1      1   \n3                lncRNA       nuORFdb&IEAtlas          2        2      2   \n4  processed_pseudogene                 False          4        4      4   \n5        protein_coding                 False          2        2      2   \n6        protein_coding               nuORFdb          2        2      2   \n7        protein_coding       nuORFdb&IEAtlas          1        1      1   \n\n   IEAtlas  gene_name_OLD  \n0       10             10  \n1        4              4  \n2        1              1  \n3        2              2  \n4        4              4  \n5        2              2  \n6        2              2  \n7        1              1  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_type</th>\n      <th>Immunoevidence_Source</th>\n      <th>gene_name</th>\n      <th>gene_id</th>\n      <th>nuORF</th>\n      <th>IEAtlas</th>\n      <th>gene_name_OLD</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>10</td>\n      <td>10</td>\n      <td>10</td>\n      <td>10</td>\n      <td>10</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>lncRNA</td>\n      <td>IEAtlas</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>lncRNA</td>\n      <td>nuORFdb</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>lncRNA</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>processed_pseudogene</td>\n      <td>False</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n      <td>4</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>protein_coding</td>\n      <td>False</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>protein_coding</td>\n      <td>nuORFdb</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>protein_coding</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidates_subset_biomartOLD.groupby([\"gene_type\",\"Immunoevidence_Source\"]).count().reset_index()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:25.614538292Z",
     "start_time": "2023-10-17T10:55:25.598325970Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "outputs": [],
   "source": [
    "raw_candidates_immunoevidence = candidates_subset_biomartOLD[['gene_name', 'gene_name_OLD','gene_type', 'Immunoevidence_Source']].merge(raw_candidates.drop(['gene_id'], axis=1), on=[\"gene_name\",\"gene_type\"])\n",
    "raw_candidates_immunoevidence.to_csv(os.path.join(GENERAL,\"RNASeq_expression/tumorspecific_3TPM_cohorts_10percent_immunoevicenceSource.csv\"), index=False)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T10:55:29.870655552Z",
     "start_time": "2023-10-17T10:55:29.853065035Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "outputs": [
    {
     "data": {
      "text/plain": "          gene_name gene_name_OLD             gene_type Immunoevidence_Source  \\\n0            CYCSP6        CYCSP6  processed_pseudogene                 False   \n1             DPPA2         DPPA2        protein_coding               nuORFdb   \n2   ENSG00000224516    AC068134.1                lncRNA                 False   \n3   ENSG00000229131    AC016710.1                lncRNA                 False   \n4   ENSG00000230105    AL354793.1                lncRNA               IEAtlas   \n5   ENSG00000234464    AL356010.2                lncRNA                 False   \n6   ENSG00000247402    AC099487.1                lncRNA                 False   \n7   ENSG00000249894    AC024581.1                lncRNA                 False   \n8   ENSG00000251550    AC091133.5  processed_pseudogene                 False   \n9   ENSG00000253154    AC100801.1                lncRNA       nuORFdb&IEAtlas   \n10  ENSG00000253474    AC044893.1                lncRNA               nuORFdb   \n11  ENSG00000255689    AP003174.1                lncRNA                 False   \n12  ENSG00000260073    AC023824.1                lncRNA       nuORFdb&IEAtlas   \n13  ENSG00000262198    AL157931.1                lncRNA                 False   \n14  ENSG00000270495    AC008825.2  processed_pseudogene                 False   \n15             EPYC          EPYC        protein_coding                 False   \n16            FGF19         FGF19        protein_coding       nuORFdb&IEAtlas   \n17         KRTAP3-1      KRTAP3-1        protein_coding                 False   \n18        LINC01456     LINC01456                lncRNA                 False   \n19        LINC02228     LINC02228                lncRNA               IEAtlas   \n20        LINC02474     LINC02474                lncRNA               IEAtlas   \n21        LINC02663     LINC02663                lncRNA                 False   \n22        LINC03004    AL356234.2                lncRNA               IEAtlas   \n23           MAGEC1        MAGEC1        protein_coding               nuORFdb   \n24        NDUFA5P10     NDUFA5P10  processed_pseudogene                 False   \n25       ZBTB46-AS1    ZBTB46-AS1                lncRNA                 False   \n\n    UC-GENOME  IMvigor210  SNY-2017  UNC-108  HdM-BLCA-1  total_n  \\\n0          28          85         6       26           8      153   \n1           3           6        15        2           0       26   \n2           1          17         0        0           1       19   \n3           3           0         0        0           0        3   \n4          26           0         0        0          12       38   \n5           0           0         1        0           0        1   \n6           9           0         0        0           0        9   \n7          11           0         0        0           9       20   \n8           4           7         2        0           0       13   \n9           2           0         0        0           1        3   \n10          4           0         0        0           1        5   \n11          2           0         0        0           0        2   \n12          0           0         0        0           0        0   \n13          4           1         0        0           2        7   \n14          3           0         0        0           0        3   \n15         10          53         6        7           3       79   \n16          4           0         2        0           0        6   \n17          4           4         5        3           2       18   \n18          6           0         0        0           0        6   \n19          1           0         0        0           0        1   \n20         22          19         3        0           6       50   \n21          4           0         0        0           1        5   \n22          0           0         0        0           0        0   \n23          1          23        12        4           3       43   \n24         17           0         4        2           1       24   \n25         21           0         3        0           0       24   \n\n    total_percentage  nuORF  Chong  vanHeesch  IEAtlas  \n0              36.52  False  False      False    False  \n1               6.21   True  False      False    False  \n2               4.53  False  False      False    False  \n3               0.72  False  False      False    False  \n4               9.07  False  False      False     True  \n5               0.24  False  False      False    False  \n6               2.15  False  False      False    False  \n7               4.77  False  False      False    False  \n8               3.10  False  False      False    False  \n9               0.72   True  False      False     True  \n10              1.19   True  False      False    False  \n11              0.48  False  False      False    False  \n12              0.00   True  False      False     True  \n13              1.67  False  False      False    False  \n14              0.72  False  False      False    False  \n15             18.85  False  False      False    False  \n16              1.43   True  False      False     True  \n17              4.30  False  False      False    False  \n18              1.43  False  False      False    False  \n19              0.24  False  False      False     True  \n20             11.93  False  False      False     True  \n21              1.19  False  False      False    False  \n22              0.00  False  False      False     True  \n23             10.26   True  False      False    False  \n24              5.73  False  False      False    False  \n25              5.73  False  False      False    False  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>gene_name_OLD</th>\n      <th>gene_type</th>\n      <th>Immunoevidence_Source</th>\n      <th>UC-GENOME</th>\n      <th>IMvigor210</th>\n      <th>SNY-2017</th>\n      <th>UNC-108</th>\n      <th>HdM-BLCA-1</th>\n      <th>total_n</th>\n      <th>total_percentage</th>\n      <th>nuORF</th>\n      <th>Chong</th>\n      <th>vanHeesch</th>\n      <th>IEAtlas</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>CYCSP6</td>\n      <td>CYCSP6</td>\n      <td>processed_pseudogene</td>\n      <td>False</td>\n      <td>28</td>\n      <td>85</td>\n      <td>6</td>\n      <td>26</td>\n      <td>8</td>\n      <td>153</td>\n      <td>36.52</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>DPPA2</td>\n      <td>DPPA2</td>\n      <td>protein_coding</td>\n      <td>nuORFdb</td>\n      <td>3</td>\n      <td>6</td>\n      <td>15</td>\n      <td>2</td>\n      <td>0</td>\n      <td>26</td>\n      <td>6.21</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENSG00000224516</td>\n      <td>AC068134.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>1</td>\n      <td>17</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>19</td>\n      <td>4.53</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENSG00000229131</td>\n      <td>AC016710.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENSG00000230105</td>\n      <td>AL354793.1</td>\n      <td>lncRNA</td>\n      <td>IEAtlas</td>\n      <td>26</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>12</td>\n      <td>38</td>\n      <td>9.07</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>ENSG00000234464</td>\n      <td>AL356010.2</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>ENSG00000247402</td>\n      <td>AC099487.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>9</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>2.15</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>ENSG00000249894</td>\n      <td>AC024581.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>11</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>9</td>\n      <td>20</td>\n      <td>4.77</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>ENSG00000251550</td>\n      <td>AC091133.5</td>\n      <td>processed_pseudogene</td>\n      <td>False</td>\n      <td>4</td>\n      <td>7</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>13</td>\n      <td>3.10</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>ENSG00000253154</td>\n      <td>AC100801.1</td>\n      <td>lncRNA</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>10</th>\n      <td>ENSG00000253474</td>\n      <td>AC044893.1</td>\n      <td>lncRNA</td>\n      <td>nuORFdb</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>11</th>\n      <td>ENSG00000255689</td>\n      <td>AP003174.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0.48</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>12</th>\n      <td>ENSG00000260073</td>\n      <td>AC023824.1</td>\n      <td>lncRNA</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>13</th>\n      <td>ENSG00000262198</td>\n      <td>AL157931.1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>4</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>2</td>\n      <td>7</td>\n      <td>1.67</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>14</th>\n      <td>ENSG00000270495</td>\n      <td>AC008825.2</td>\n      <td>processed_pseudogene</td>\n      <td>False</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0.72</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>EPYC</td>\n      <td>EPYC</td>\n      <td>protein_coding</td>\n      <td>False</td>\n      <td>10</td>\n      <td>53</td>\n      <td>6</td>\n      <td>7</td>\n      <td>3</td>\n      <td>79</td>\n      <td>18.85</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>FGF19</td>\n      <td>FGF19</td>\n      <td>protein_coding</td>\n      <td>nuORFdb&amp;IEAtlas</td>\n      <td>4</td>\n      <td>0</td>\n      <td>2</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>17</th>\n      <td>KRTAP3-1</td>\n      <td>KRTAP3-1</td>\n      <td>protein_coding</td>\n      <td>False</td>\n      <td>4</td>\n      <td>4</td>\n      <td>5</td>\n      <td>3</td>\n      <td>2</td>\n      <td>18</td>\n      <td>4.30</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>18</th>\n      <td>LINC01456</td>\n      <td>LINC01456</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>6</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>6</td>\n      <td>1.43</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>19</th>\n      <td>LINC02228</td>\n      <td>LINC02228</td>\n      <td>lncRNA</td>\n      <td>IEAtlas</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0.24</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>20</th>\n      <td>LINC02474</td>\n      <td>LINC02474</td>\n      <td>lncRNA</td>\n      <td>IEAtlas</td>\n      <td>22</td>\n      <td>19</td>\n      <td>3</td>\n      <td>0</td>\n      <td>6</td>\n      <td>50</td>\n      <td>11.93</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>21</th>\n      <td>LINC02663</td>\n      <td>LINC02663</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>4</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>5</td>\n      <td>1.19</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>22</th>\n      <td>LINC03004</td>\n      <td>AL356234.2</td>\n      <td>lncRNA</td>\n      <td>IEAtlas</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0.00</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>23</th>\n      <td>MAGEC1</td>\n      <td>MAGEC1</td>\n      <td>protein_coding</td>\n      <td>nuORFdb</td>\n      <td>1</td>\n      <td>23</td>\n      <td>12</td>\n      <td>4</td>\n      <td>3</td>\n      <td>43</td>\n      <td>10.26</td>\n      <td>True</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>NDUFA5P10</td>\n      <td>NDUFA5P10</td>\n      <td>processed_pseudogene</td>\n      <td>False</td>\n      <td>17</td>\n      <td>0</td>\n      <td>4</td>\n      <td>2</td>\n      <td>1</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>25</th>\n      <td>ZBTB46-AS1</td>\n      <td>ZBTB46-AS1</td>\n      <td>lncRNA</td>\n      <td>False</td>\n      <td>21</td>\n      <td>0</td>\n      <td>3</td>\n      <td>0</td>\n      <td>0</td>\n      <td>24</td>\n      <td>5.73</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_candidates_immunoevidence"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-10-17T13:03:17.525671490Z",
     "start_time": "2023-10-17T13:03:17.515177683Z"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
